### PR TITLE
Issue #258: Make connector steps execute after flow steps have finished.

### LINKED ISSUE #258: Connector steps not "executing" after writer step
Sample workflow to test on:

[Issue_#258_20220704_0542.yml.txt](https://github.com/catalyst/moodle-tool_dataflows/files/9036635/Issue_.258_20220704_0542.yml.txt)

Instructions:
1. Import workflow
2. Open `admin/tool/dataflows/classes/local/step/connector_debugging.php`
3. Add some echo, mtrace, or kill the script somewhere inside the `execute` method

Expected results: It does whatever you tell it to once it reaches the debugging step.

Actual results: It proceeds to finalise and doesn't execute the method at all.

Run output:
```
2022 Jul  4, 15:44:59.023 Engine "Issue #258": status: new 
2022 Jul  4, 15:44:59.034 Engine "Issue #258", step "SQL reader": status: new 
2022 Jul  4, 15:44:59.034 Engine "Issue #258", step "write to csv": status: new 
2022 Jul  4, 15:44:59.035 Engine "Issue #258", step "debug": status: new 
2022 Jul  4, 15:44:59.036 Engine "Issue #258", step "SQL reader": status: initialised 
2022 Jul  4, 15:44:59.036 Engine "Issue #258", step "write to csv": status: initialised 
2022 Jul  4, 15:44:59.037 Engine "Issue #258", step "debug": status: initialised 
2022 Jul  4, 15:44:59.037 Engine "Issue #258": status: initialised, config: {"isdryrun":false} 
2022 Jul  4, 15:44:59.055 Engine "Issue #258": status: processing 
2022 Jul  4, 15:44:59.055 here2 
2022 Jul  4, 15:44:59.055 Engine "Issue #258", step "debug": status: blocked 
2022 Jul  4, 15:44:59.067 Engine "Issue #258", step "write to csv": status: waiting 
2022 Jul  4, 15:44:59.094 Engine "Issue #258", step "SQL reader": status: flowing 
2022 Jul  4, 15:44:59.120 Engine "Issue #258", step "write to csv": status: flowing 
2022 Jul  4, 15:44:59.135 Engine "Issue #258": status: finished 
2022 Jul  4, 15:44:59.148 Engine "Issue #258", step "SQL reader": status: finalised 
2022 Jul  4, 15:44:59.148 Engine "Issue #258", step "write to csv": status: finalised 
2022 Jul  4, 15:44:59.148 Engine "Issue #258", step "debug": status: finalised 
2022 Jul  4, 15:44:59.149 Engine "Issue #258": status: finalised 
2022 Jul  4, 15:44:59.158 Engine "Issue #258": dumping state..
global: {  }
env:
  DATAFLOW_ID: 5
  DATAFLOW_RUN_ID: 25
  DATAFLOW_RUN_NAME: '1'
dataflow:
  name: 'Issue #258'
  config: {  }
  states:
    new: 1656913499.0233
    initialised: 1656913499.0367
    processing: 1656913499.0549
    finished: 1656913499.1345
    finalised: 1656913499.1485
steps:
  sql:
    name: 'SQL reader'
    type: tool_dataflows\local\step\reader_sql
    config:
      sql: |
        SELECT generate_series AS num
        FROM generate_series(1,4)
        
      counterfield: num
      countervalue: '4'
    alias: sql
    states:
      new: 1656913499.034
      initialised: 1656913499.0363
      flowing: 1656913499.0941
      finalised: 1656913499.1478
  writer:
    name: 'write to csv'
    depends_on: sql
    type: tool_dataflows\local\step\writer_stream
    config:
      streamname: out.csv
      format: csv
    alias: writer
    states:
      new: 1656913499.0343
      initialised: 1656913499.0365
      waiting: 1656913499.0672
      flowing: 1656913499.1203
      finalised: 1656913499.1482
  debug:
    name: debug
    depends_on: writer
    type: tool_dataflows\local\step\connector_debugging
    alias: debug
    states:
      new: 1656913499.0346
      initialised: 1656913499.0366
      blocked: 1656913499.0552
      finalised: 1656913499.1484
    config: {  }

```

### PR DESCRIPTION
**Pull request by jaypha**
Resolves #258 

This should fix both problems mentioned in the issue notes.

### CODE REVIEW

File: classes/local/execution/engine.php [hunk 1]
```diff
@@ -233,6 +249,16 @@ public function execute() {
         $this->finalise();
     }
 
+    /**
+     * Adds an engine step to the queues for processing.
+     *
+     * @param engine_step $step
+     */
+    public function add_to_queue(engine_step $step) {
+        assert(in_array($step, $this->enginesteps) || in_array($step, $this->flowcaps));
```
Comment (keevan) [inline classes/local/execution/engine.php:258]: What was the reason this assert is added here? Shouldn't it be a if throw instead?

Comment (jaypha) [reply classes/local/execution/engine.php:258]: asserts don't get compiled in production environments.

**Review by keevan — COMMENTED** [review]
Small assert thing, and a concern for the `on_change_status` function which sounds like a hook but isn't used as such.

Perhaps it could have a better name. In terms of testing it seems okay. It does fix the issue.

File: classes/local/execution/engine.php (hunk 1 above)
Comment (keevan) [reply classes/local/execution/engine.php:258]: Could you please change this to an if throw, and add a supporting comment as to why the check is needed.

**Comment by jaypha** [comment]
> Small assert thing, and a concern for the on_change_status function which sounds like a hook but isn't used as such.

What do you think it should be called?

**Comment by keevan** [comment]
> > Small assert thing, and a concern for the on_change_status function which sounds like a hook but isn't used as such.
> 
> What do you think it should be called?

adjust_streams? adjust_queue? The main functionality I can see is it adding to "the queue" based on a combination of status and whether or not it's a flow.

**Comment by jaypha** [comment]
> adjust_streams? adjust_queue? The main functionality I can see is it adding to "the queue" based on a combination of status and whether or not it's a flow.

For now. It could be different in the future, or be overridden in another class. The one thing that it definitely is, is that it is called in response to a change in status.

**Review by keevan — APPROVED** [review]

**Pull request merged**